README.md
XFile - a data management and layout tool

1. This is very preliminary README

2. The working/project directory is hard-coded as C:\xfile.  To work
properly, run xfile.exe from C:\xfile

4. Things that are there but not implemented:
  - Field notes
  - Default Values

XFile is based on ideas found in GeoFile for PC-GEOS, and FileMaker Pro 2.0
for Windows 3.1.  This document is a very brief overview of features.

- Multiple data layouts
- Image data type
- Static images and labels
- Background image
- Summary data type (Sum, Average, Max, or Min of an Integer or Real field)
- Sorting
- Marking (filtering) records by formula
- Tearoff menus (This is pretty cool, actually)

** Creating a Project ***

1. Launch XFile
2. Chose Create Project
3. Enter a project name (probably a single word)
4. Click the Design Mode button in lower left corner
5. Click Create Fields
6. Add fields of various types.
7. Switch to Data Entry mode
8. Enter data
9. Create more layouts if you want

*** Field Types ***

String (single and multi-line). Maximum of 512 characters (this is actually arbitrary, it could be anything).
Integer : Min/Max is minimum and maximum allowed value
Real    : Min/Max is minimum and maximum allowed value*
Date    : Min/Max has no effect
Summary : Select summary type, then select summary field
Image   : Min/Max are fixed width and height
      (if supplied) otherwise the field resizes to hold the supplied bitmap

Fields may be freely placed anywhere within the "canvas", but should not overlap.
I have not seen any problems with overlapping fields, but XBasic recommends against it.
It is possible (though difficult) to move a field or label outside the bounds of the canvas,
and be irretrievable.  This is a known issue.

The Image field type:
- When in Data Entry mode, left click on the image field
- In the supplied file navigation dialog, select an image file 
- Supported image formats: BMP, JPG, PNG, GIF, TIF, PCX (uses file extension to determine format)
- The selected image will be converted to a 24-bit bitmap if needed.
- The selected bitmap will be copied to the /images directory of your project.
- If the image was converted to a bitmap, the .bmp is moved to your images directory.
- The image name will be stored in the database for that record.
- The image name cannot exceed 32 characters.

Selection Mode:
- Click the little selection button on the left
- Group selection with click and drag
- Single selection with left mouse
- Add another selection with Ctrl-Left Mouse
- (shift) Tab key and middle mouse button cycles through single-selection

Resize:
- Fields can be resized.
- Select a field (objects like labels and static images are not resizable)
- Use Ctrl + arrow keys to resize.  Hold down shift to resize by 5 pixels.

Move:
- Use the mouse to select and move fields
- Also, use arrow keys to move selected fields
- Make use of the Arrange tools under Layout menu in Design mode

Sample Databases
-comics
-todo_list
-photos

src/CHANGELOG
3/10/2009
- Added Delete to Edit menu of Design Mode
- Added group delete (only applies to Labels/Static Images)
- Added Select All to Edit Menu of Design Mode
- Fixed delete record handling.  It now picks the right
  next record to display after a record is deleted.
- Added ability to change the order of the fields in the database.
- Summary fields obey the Show Only Marked rule.  They will operate
  only on marked records when Show Only Marked is true.
- Only update fields that are in the current layout

3/12/2009
- Fixed Field Order code
- Changed font application-wide
- Changed some XuiList scrollbars
- Added Field Size indicator to Ruler
- Added labels to pixel mode of Ruler
- Changed Ruler to green

3/13/2009
- Fixed problem with Reload Current layout
- Always show field names in Design mode
- More font changes
- fix for smearing when in "Show Window Contents While Dragging" mode.
- Added QuitProgram() to handle quitting properly 

3/14/2009
- Save button saves Layout in Design mode, Records in Data mode
- Open and New buttons only operate in Design Mode
- Switching modes keeps main window selected
- You can resize fields!  Arrow keys on selected field

3/15/2009
- Added sanity checks to move and resize
- You can move with the arrow keys. (Shift to move or resize by a greater increment)
- You can resize with Ctrl-arrow keys
- Size guide shows up in ruler for a selected field (or first selected field in the case of a group select)
- FieldProperties window updates font viewer to show font of selected object.
- Cannot create a pages size that wont fit on the screen
 
3/17/2009
- Only update fields in the current layout when displaying a record
- Only redraw Canvas when needed in SetImage
- Unmarking all records turns off Show Only Marked

3/18/2009
- Added Export dialog.  Allows export as CSV or BMP
- Added ExpressionBuilder.  Allows some better marking options.  Allows
  a query like [field] [,>,<,>=,<=] [value] [AND/OR] [field] [,>,<,>=,<=] [value]

3/20/2009
- Added support for inserting date and time
- Fixed issue with tearoffs not being right size
- Got rid of TearoffInit
- Allowed CloseProject() to bail out if field validation fails.

3/21/2009
- Deleting a field now triggers summary field updates
- Got rid of unused component in FIELD type definition (breaks all existing layouts)
- Viewing summary field from Organizer now shows Data Attributes properly
- Deleting a field now updates Summary field index references.
- Re-ordering fields updates Summary field index references.
- Fix minute value for Insert Time if minute is < 10 (adds the leading 0)

3/23/2009
- Fixed CSV export/import so XFile can at least import its own exported files. (see README)
- Added Insert | Week Of Year.  Now you dont have to waste both of your expression in ExpressionBuilder
  to calculated dates falling within the current week.  You can have an integer field that holds 
  week of year instead.

3/24/2009
- XFile window title now properly displays current project name
- Fixed small issue with Select All
- Fixed small issue with Fonts during layout load
- Fixed issue with operator lists being horizontally scrollable in ExpressionBuilder

3/25/2009
- Fixed issue in Create dialog
- Made Length field manually editable
- Added preliminary support (via PVW32Conv.exe) for JPG, PNG, GIF, TIF, PCX images

3/27/2009
- Fixed boundary check for mouse move of objects
- Got rid of unused function
- Cleaned up Select mode mouse handling
- Mouse wheel can now be used to navigate records in Data mode
- Mouse wheel now acts like Tab key in Design mode

3/28/2009
- Allow blank date

3/29/2009
- Change radio button style on FieldProperties dialog
- Added xfile.cfg.  Sets a few global configuration parameters.  Editable/re-readable from within XFile (Properties menu)

3/30/2009
- Added project title to Console window
- Added parameters to xfile.cfg for Debug and ShowConsole
- Added Show Console and Hide Console to both View menus
- Intelligent tab key handling in Data Entry mode (still goofy in Design Mode)

3/31/2009
- Intelligent tab key handling extended to Design Mode.  Ugly, ugly code.
- Removed Bitmap Editor from .geo file.  It is now read only from xfile.cfg.
- Fixed handling of image names with spaces in them
- Fixed handling of file extension
- Fixed issue where ruler guides would not redraw if mouse was not
  in Canvas grid

4/4/2009
- Added '-' to disallowed characters in field names

6/10/2009
- Added 'Quit' menu option to File menu in Data mode
- Unmarking records in "Show Only Marked" mode selects next marked record

8/20/2009
- Added View->Field As List in Data Entry mode
  Allows viewing of all values in the database for the specified
  field in list form.  Double clicking a list entry displays the
  associated record.

8/21/2009
- Fixed Field As List to behave properly with sorting

9/5/2009
- Fixed crash if you type a field name in View Field As List that does not exist.
- Added simple table based HTML export. Man, that export code is a mess.
- Added View Database As HTML, which runs the HTML export to a temp file, then
  loads the file in the HTMLViewer program defined in the config file.  So, you
  will want to add a line HTMLViewer = "whatever program" to your xfile.cfg file.
  A neat thing is that Excel, and probably any other modern spreadsheet will import
  the HTML nicely into a spreadsheet.
- If you are in design mode, and chose Run Bitmap Editor, if you have a static image 
  selected, that static image will be passed as the first argument to your bitmap editor.
  If you modify the image, you can select Layout | Reload Current to see the changes.
  
** PENDING KNOWN ISSUES (incomplete list)
2. The multiline text field doesnt resize automatically when the font is changed
4. On occassion, the ruler draws beyond its bounds.

src/tokenize.s
;
; PROGRAM "tokenize"
;
;
; DECLARE FUNCTION  Entry ()
.data
.align 8
%%%firstStatic:
%%%entered:
.zero  8
;
.text
.globl _Entry@0
;
;
; EXPORT
;
; DECLARE FUNCTION  Tokenize (in$, @out$[], @delimiter$)
.globl _Tokenize@12
;
; END EXPORT
;
;
; FUNCTION  Entry ()
.text
.text
	jmp	%_StartLibrary_tokenize		;;; 200
PrologCode:
	ret			;;; 202
;
.globl  %_StartLibrary_tokenize
%_StartLibrary_tokenize:
	call	func1		;;; 208
	ret	0		;;; 209
;
_Entry@0:
;
;  *****
;  *****  FUNCTION  Entry ()  *****
;  *****
func1:
	push	ebp		;;; 110
	mov	ebp,esp		;;; 111
	sub	esp,256		;;; 112
	mov	[ebp-12],esi		;;; 113
	mov	[ebp-16],edi		;;; 114
	mov	[ebp-20],ebx		;;; 115
	call	%%%%initOnce		;;; 116
;
funcBody1:
;
; IF LIBRARY(0) THEN RETURN
	sub	esp,64		;;; 627
	mov	eax,0		;;; 868
	xor	eax,eax		;;; 682
	not	eax		;;; 683
	add	esp,64		;;; 809
	or	eax,eax		;;; 269
	jz	else.0001		;;; 270
	xor	eax,eax		;;; 1108
	jmp	end.func1		;;; 313
else.0001:
end.if.0001:
;
; END FUNCTION
	xor	eax,eax		;;; 1108
;;
end.func1:
	mov	esi,[ebp-12]		;;; 100
	mov	edi,[ebp-16]		;;; 101
	mov	ebx,[ebp-20]		;;; 102
	mov	esp,ebp		;;; 103
	pop	ebp		;;; 104
	ret			;;; 108
;
;
%%%%initOnce:
	mov	eax,[%%%entered]		;;; 127
	or	eax,eax		;;; 128
	jnz	%%%%initOnceDone		;;; 129
	mov	esi,%%%firstStatic		;;; 130
	mov	edi,%%%lastStatic		;;; 131
	call	%_ZeroMemory		;;; 137
	call	PrologCode		;;; 138
	call	InitSharedComposites		;;; 139
	mov	[%%%entered],-1		;;; 140
%%%%initOnceDone:
	ret			;;; 142
;
;
.globl %%%%blowback_tokenize
%%%%blowback_tokenize:
	mov	[%%%entered],0		;;; 143
	ret			;;; 144
;
;
; FUNCTION  Tokenize (@in$, @out$[], @delim$)
.text
_Tokenize@12:
;
;  *****
;  *****  FUNCTION  Tokenize ()  *****
;  *****
func2:
	push	ebp		;;; 110
	mov	ebp,esp		;;; 111
	sub	esp,256		;;; 112
	mov	[ebp-12],esi		;;; 113
	mov	[ebp-16],edi		;;; 114
	mov	[ebp-20],ebx		;;; 115
	cld			;;; 117
	lea	edi,[ebp-44]		;;; 118
	mov	ecx,6		;;; 119
	xor	eax,eax		;;; 120
	rep			;;; 121
	stosd			;;; 122
;
funcBody2:
;
; XLONG i
;
; XLONG count
;
; st$ = ""
	xor	eax,eax		;;; 3
	lea	ebx,[ebp-32]		;;; 6
	mov	esi,[ebx]		;;; 7
	mov	[ebx],eax		;;; 8
	call	%____free		;;; 9
;
;
; REDIM out$[]
	sub	esp,64		;;; 536
	mov	esi,[ebp+12]		;;; 875
	call	%_FreeArray		;;; 537
	mov	esi,0		;;; 877
	mov	[ebp+12],esi		;;; 884
	add	esp,64		;;; 539
;
; count = 0
	mov	eax,0		;;; 868
	mov	[ebp-28],eax		;;; 884
;
;
; IF(in$ == "") THEN RETURN 0
	mov	eax,[ebp+8]		;;; 875
	xor	ebx,ebx		;;; 867
	call	%_string.compare.vv		;;; 909
	jne	else.0002		;;; 268
	mov	eax,0		;;; 868
	jmp	end.func2		;;; 313
else.0002:
end.if.0002:
;
;
; IF(delim$ == "") THEN
	mov	eax,[ebp+16]		;;; 875
	xor	ebx,ebx		;;; 867
	call	%_string.compare.vv		;;; 909
	jne	else.0003		;;; 268
;
; REDIM out$[0] : out$[0] = in$
	sub	esp,64		;;; 536
	mov	eax,0		;;; 868
	mov	[esp+16],eax		;;; 540
	mov	esi,[ebp+12]		;;; 875
	mov	[esp],esi		;;; 541
	mov	[esp+4],1		;;; 542
	mov	[esp+8],-1072496636		;;; 543
	mov	[esp+12],0		;;; 544
	call	%_RedimArray		;;; 545
	add	esp,64		;;; 546
	mov	[ebp+12],eax		;;; 884
	mov	eax,[ebp+8]		;;; 875
	call	%_clone.a0		;;; 30
	mov	ecx,[ebp+12]		;;; 875
	lea	ebx,[ecx]		;;; 581
	mov	esi,[ebx]		;;; 36
	mov	[ebx],eax		;;; 37
	call	%____free		;;; 38
;
; RETURN 1
	mov	eax,1		;;; 868
	jmp	end.func2		;;; 313
;
; END IF
else.0003:
end.if.0003:
;
;
; FOR i = 0 TO LEN(in$)
	mov	eax,0		;;; 868
	mov	[ebp-24],eax		;;; 884
	mov	eax,[ebp+8]		;;; 875
	or	eax,eax		;;; 793
	jz	%0003		;;; 794
	mov	eax,[eax-8]		;;; 795
%0003:
	mov	[ebp-36],eax		;;; 884
for.0004:
	mov	eax,[ebp-24]		;;; 875
	mov	ebx,[ebp-36]		;;; 875
	cmp	eax,ebx		;;; 194
	jg	end.for.0004		;;; 195
;
; IFZ(INSTR(delim$, CHR$(in${i}))) THEN
	sub	esp,64		;;; 627
	mov	eax,[ebp+16]		;;; 875
	mov	[esp],eax		;;; 1130
	sub	esp,64		;;; 627
	mov	eax,[ebp-24]		;;; 875
	mov	edx,[ebp+8]		;;; 875
	movzx	eax,byte ptr [edx+eax]		;;; 581
	mov	[esp],eax		;;; 1130
	mov	[esp+4],1		;;; 757
	call	%_chr.d		;;; 758
	add	esp,64		;;; 809
	mov	[esp+4],eax		;;; 1130
	mov	[esp+8],0		;;; 749
	call	%_instr.vs		;;; 750
	add	esp,64		;;; 809
	or	eax,eax		;;; 241
	jnz	else.0005		;;; 242
;
; IF(i == LEN(in$)) THEN
	mov	eax,[ebp+8]		;;; 875
	or	eax,eax		;;; 793
	jz	%0004		;;; 794
	mov	eax,[eax-8]		;;; 795
%0004:
	mov	ebx,[ebp-24]		;;; 875
	cmp	eax,ebx		;;; 902
	jne	else.0006		;;; 268
;
; GOSUB Add
	call	%s%Add%2		;;; 211
;
; RETURN count
	mov	eax,[ebp-28]		;;; 875
	jmp	end.func2		;;; 313
;
; END IF
else.0006:
end.if.0006:
;
; st$ = st$ + CHR$(in${i})
	sub	esp,64		;;; 627
	mov	eax,[ebp-24]		;;; 875
	mov	edx,[ebp+8]		;;; 875
	movzx	eax,byte ptr [edx+eax]		;;; 581
	mov	[esp],eax		;;; 1130
	mov	[esp+4],1		;;; 757
	call	%_chr.d		;;; 758
	add	esp,64		;;; 809
	mov	ebx,[ebp-32]		;;; 875
	call	%_concat.ubyte.a0.eq.a1.plus.a0.vs		;;; 992
	lea	ebx,[ebp-32]		;;; 6
	mov	esi,[ebx]		;;; 7
	mov	[ebx],eax		;;; 8
	call	%____free		;;; 9
;
; ELSE
	jmp	end.if.0005		;;; 98
else.0005:
;
; GOSUB Add
	call	%s%Add%2		;;; 211
;
; END IF
end.if.0005:
;
; NEXT i
do.next.0004:
	inc	[ebp-24]		;;; 278
	jmp	for.0004		;;; 280
end.for.0004:
;
; RETURN count
	mov	eax,[ebp-28]		;;; 875
	jmp	end.func2		;;; 313
;
;
; SUB Add
	jmp	out.sub2.0		;;; 317
%s%Add%2:
;
; IF(st$ != "") THEN
	mov	eax,[ebp-32]		;;; 875
	xor	ebx,ebx		;;; 867
	call	%_string.compare.vv		;;; 909
	je	else.0007		;;; 268
;
; size = UBOUND(out$[])
	mov	eax,[ebp+12]		;;; 875
	or	eax,eax		;;; 793
	jz	%0006		;;; 794
	mov	eax,[eax-8]		;;; 795
%0006:
	dec	eax		;;; 792
	mov	[ebp-44],eax		;;; 884
;
; REDIM out$[size + 1]
	sub	esp,64		;;; 536
	mov	eax,[ebp-44]		;;; 875
	add	eax,1		;;; 986
	mov	[esp+16],eax		;;; 540
	mov	esi,[ebp+12]		;;; 875
	mov	[esp],esi		;;; 541
	mov	[esp+4],1		;;; 542
	mov	[esp+8],-1072496636		;;; 543
	mov	[esp+12],0		;;; 544
	call	%_RedimArray		;;; 545
	add	esp,64		;;; 546
	mov	[ebp+12],eax		;;; 884
;
; out$[UBOUND(out$[])] = st$
	mov	eax,[ebp-32]		;;; 875
	call	%_clone.a0		;;; 30
	mov	ebx,[ebp+12]		;;; 875
	or	ebx,ebx		;;; 793
	jz	%0007		;;; 794
	mov	ebx,[ebx-8]		;;; 795
%0007:
	dec	ebx		;;; 792
	mov	ecx,[ebp+12]		;;; 875
	lea	ebx,[ecx+ebx*4]		;;; 581
	mov	esi,[ebx]		;;; 36
	mov	[ebx],eax		;;; 37
	call	%____free		;;; 38
;
; INC count
	inc	[ebp-28]		;;; 67
;
; st$ = ""
	xor	eax,eax		;;; 3
	lea	ebx,[ebp-32]		;;; 6
	mov	esi,[ebx]		;;; 7
	mov	[ebx],eax		;;; 8
	call	%____free		;;; 9
;
; END IF
else.0007:
end.if.0007:
;
; END SUB
end.sub2.0:
	ret			;;; 145
out.sub2.0:
;
; END FUNCTION
	xor	eax,eax		;;; 1108
;;
end.func2:
	mov	esi,[ebp-32]		;;; 875
	call	%____free		;;; 531
	xor	edx,edx		;;; 532
	mov	[ebp-32],edx		;;; 884
	mov	esi,[ebp-12]		;;; 100
	mov	edi,[ebp-16]		;;; 101
	mov	ebx,[ebp-20]		;;; 102
	mov	esp,ebp		;;; 103
	pop	ebp		;;; 104
	ret	12		;;; 105
;
; END PROGRAM
end_program:
.data
.zero  8
%%%lastStatic:
.zero  8
.text
	push	ebp		;;; 146
	mov	ebp,esp		;;; 147
	sub	esp,128		;;; 148
;
;
;
;
	mov	esp,ebp		;;; 151
	pop	ebp		;;; 152
	ret			;;; 153
;
;
InitSharedComposites:
	ret			;;; 163
;
;
; *****  DEFINE LITERAL STRINGS  *****
;
.text
.align 8
.dword	32, 0, 8, 0x80130001
@_string.0029:
.byte	"tokenize"
.zero	8
.dword	32, 0, 5, 0x80130001
@_string.Entry:
.byte	"Entry"
.zero	11
.dword	32, 0, 15, 0x80130001
@_string.StartLibrary:
.byte	"%_StartLibrary_"
.zero	1
